### PR TITLE
IR: do not create unused capture fields

### PR DESCRIPTION
**Pull request by pyos**
I.e. those only used in constructors, where parameters with the same values are also present.

One exception is captured inline parameters on the JVM backend, as the bytecode inliner uses field assignment instructions (setfield) to locate them; removing the field is thus not possible:

```
inline fun baz(crossinline f: () -> String) =
    object {
        init { println(f()) } // *has* to create a field called `$f`...
    }

fun main() {
    val ok = "OK"
    println(baz { ok }) // ...otherwise the inliner won't create a field for `ok`.
}
```

TODO: fix the inliner? Even so, removing the field would be a backwards-incompatible change.

^KT-48784 Fixed

### CODE REVIEW

File: compiler/testData/codegen/bytecodeListing/annotations/localClassWithCapturedParams_ir.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+@kotlin.Metadata
```
Comment (pyos) [inline compiler/testData/codegen/bytecodeListing/annotations/localClassWithCapturedParams_ir.txt:1]: Diff in this test:
```
annotation class Simple(val value: String)

fun localCaptured(): Any {
    val z  = 1
    class A(@Simple("K") val z: String) {
        val x = z // no longer creates `final synthetic String $z`
    }
    return A("K")
}
```
The fact that `z` refers to the local `z` and not the property `z` looks questionable to me, but that's not the point right now.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt [hunk 2]
```diff
@@ -163,36 +165,55 @@ class LocalDeclarationsLowering(
...
+        val declaration: IrClass,
+        val inInlineFunctionScope: Boolean,
+        val constructorContext: LocalContext?
+    ) : LocalContext() {
         lateinit var closure: Closure
 
         // NOTE: This map is iterated over in `rewriteClassMembers` and we're relying on
         // the deterministic iteration order that `mutableMapOf` provides.
-        val capturedValueToField: MutableMap<IrValueDeclaration, IrField> = mutableMapOf()
+        val capturedValueToField: MutableMap<IrValueDeclaration, Lazy<IrField>> = mutableMapOf()
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt:177]: This is minor, but maybe make it a bit more explicit, by using a custom class instead of `Lazy`? It'll make it easier to inspect where exactly the related logic is (searching for usages of Lazy is costly). E.g. (not sure about the names)
```
    private class FieldToBeGenerated(generate: () -> IrField) {
        private var generate = lazy(LazyThreadSafetyMode.NONE, generate)

        fun get(): IrField = generate.value

        val isAccessed: Boolean get() = generate.isInitialized()
    }
```

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt [hunk 3]
```diff
@@ -163,36 +165,55 @@ class LocalDeclarationsLowering(
...
 
         override fun irGet(startOffset: Int, endOffset: Int, valueDeclaration: IrValueDeclaration): IrExpression? {
-            val field = capturedValueToField[valueDeclaration] ?: return null
+            // On the JVM backend, `AnonymousObjectTransformer` in the bytecode inliner uses field assignment
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt:180]: Maybe add a boxInline test?

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt [hunk 4]
```diff
@@ -953,7 +974,15 @@ class LocalDeclarationsLowering(
                     // LDL doesn't work on the enums because local enums are not allowed, so skipping them in scripts too
                     if (isScriptMode && declaration.kind == ClassKind.ENUM_CLASS) return
 
-                    localClasses[declaration] = LocalClassContext(declaration, data.inInlineFunctionScope)
+                    // If there are many primary constructors, each copy of the initializer requires different remapping.
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt:977]: I'm out of the loop... How is it possible to have many primary constructors in the same class? O_o

Comment (gogabr) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt:977]: An even more naive question: why not use `IrConstructor.isPrimary`?

Comment (pyos) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt:977]: Not sure if calling them "primary" is correct, but:
```
class C {
    constructor() {}
    constructor(x: Int) {}
    val y = something // copied into both constructors
}
```
This is why InitializersLowering does deep-copying, and why JvmLocalClassPopupLowering is a thing at all.

The `callsSuper` stuff is borrowed from [another part of LocalDeclarationsLowering](https://github.com/JetBrains/kotlin/blob/0864f9f/compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt#L473-L477). `isPrimary` (or `primaryConstructor`) didn't work for some reason for some edge case, but I don't remember what it was, I'll rerun the tests tomorrow. Maybe instead of `callsSuper` both places should check for an `IrInstanceInitializerCall` though.

Comment (pyos) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt:977]: `localClassConstructors[declaration.primaryConstructor]` *does* seem to work so I don't know what I did wrong previously, but obviously it doesn't optimize the above code even if it only has one constructor.

As for `IrInstanceInitializerCall`, looks like sometimes we forget to add it in other lowerings...

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt (hunk 2 above)
Comment (pyos) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt]: Done. Rather than passing a function it's probably better to only have a symbol, then create IrFields for the used ones later.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt (hunk 4 above)
Comment (pyos) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/LocalDeclarationsLowering.kt]: Ok, "sometimes" means only in SuspendLambdaLowering.

**Review by gogabr — APPROVED** [review]

**Comment by udalov** [comment]
Could you please fix compilation of kotlin-native in https://github.com/pyos/kotlin/blob/87b2e6e67aa94f5359d638ed269d0cbd612353d4/kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InnerClassLowering.kt#L155?

**Comment by pyos** [comment]
Done, I think? Weird that IDEA doesn't analyze this module.

**Comment by udalov** [comment]
Native modules are excluded by default, and are only analyzed/compiled if you have `kotlin.native.enabled=true` in `local.properties`. More info at `kotlin-native/README.md`.

**Comment by udalov** [comment]
Unfortunately it makes native tests fail:
```
> Task :kotlin-native:backend.native:tests:AbstractMutableCollection FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':kotlin-native:backend.native:tests:AbstractMutableCollection'.
> Test failed. Expected exit status: 0, actual: 139
  stdout:
  
  stderr:
```

**Comment by pyos** [comment]
Ah, I guess that's why the complicated way of finding non-delegated constructors. JVM & JS run LocalDeclarationsLowering before InitializersLowering, but Native runs it directly after.

**Comment by pyos** [comment]
Rolled that commit back, it wasn't important anyway (and ideally JVM/JS should run LocalDeclarationsLowering after InitializersLowering too...)

**Comment by udalov** [comment]
Thank you!
